### PR TITLE
Cross-Region Copying Pre-Signer for DocDB and Neptune

### PR DESCRIPTION
**Pull request by joviegas**
## Description

Cross-Region Copying Pre-Signer for DocDB and Neptune

## Motivation and Context

**Neptune** (As of today, `PreSignedUrl` for both operations are not currently supported)

* `CopyDBClusterSnapshot`
* `CreateDBCluster`

**DocDB** (As of today, `PreSignedUrl` for both operations are not currently supported)

* `CopyDBClusterSnapshot`
* `CreateDBCluster`

## Testing

- JUnits Added

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have read the **README** document
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [ ] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: services/docdb/src/main/java/software/amazon/awssdk/services/docdb/internal/execution.interceptors [hunk 1]
```diff
@@ -0,0 +1,2 @@
+software.amazon.awssdk.services.docdb.internal.CopyDbClusterSnapshotPresignInterceptor
```
Comment (Quanzzzz) [inline services/docdb/src/main/java/software/amazon/awssdk/services/docdb/internal/execution.interceptors:1]: Are these two imports needed? What is this class for?

File: services/docdb/src/main/resources/software/amazon/awssdk/services/docdb/execution.interceptors [hunk 2]
```diff
@@ -0,0 +1,2 @@
+software.amazon.awssdk.services.docdb.internal.CopyDbClusterSnapshotPresignInterceptor
```
Comment (Quanzzzz) [inline services/docdb/src/main/resources/software/amazon/awssdk/services/docdb/execution.interceptors:1]: Same question as above.

File: services/neptune/src/main/java/software/amazon/awssdk/services/neptune/internal/execution.interceptors [hunk 3]
```diff
@@ -0,0 +1,2 @@
+software.amazon.awssdk.services.neptune.internal.CopyDbClusterSnapshotPresignInterceptor
```
Comment (Quanzzzz) [inline services/neptune/src/main/java/software/amazon/awssdk/services/neptune/internal/execution.interceptors:1]: Same question as above.

File: services/neptune/src/main/resources/software/amazon/awssdk/services/neptune/execution.interceptors [hunk 4]
```diff
@@ -0,0 +1,2 @@
+software.amazon.awssdk.services.neptune.internal.CopyDbClusterSnapshotPresignInterceptor
```
Comment (Quanzzzz) [inline services/neptune/src/main/resources/software/amazon/awssdk/services/neptune/execution.interceptors:1]: Same question as above.

Comment (joviegas) [reply services/neptune/src/main/resources/software/amazon/awssdk/services/neptune/execution.interceptors:1]: This is required

File: services/docdb/src/main/resources/software/amazon/awssdk/services/docdb/execution.interceptors (hunk 2 above)
Comment (joviegas) [reply services/docdb/src/main/resources/software/amazon/awssdk/services/docdb/execution.interceptors:1]: This is also required

File: services/docdb/src/main/java/software/amazon/awssdk/services/docdb/internal/execution.interceptors (hunk 1 above)
Comment (joviegas) [reply services/docdb/src/main/java/software/amazon/awssdk/services/docdb/internal/execution.interceptors]: This was accidental..removed this

File: services/neptune/src/main/java/software/amazon/awssdk/services/neptune/internal/execution.interceptors (hunk 3 above)
Comment (joviegas) [reply services/neptune/src/main/java/software/amazon/awssdk/services/neptune/internal/execution.interceptors]: This was accidental..removed this

File: services/neptune/src/test/java/software/amazon/awssdk/services/neptune/internal/PresignRequestHandlerTest.java [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+                "&TargetDBClusterSnapshotIdentifier=test-instance-ss-copy-2" +
+                "&KmsKeyId=arn%3Aaws%3Akms%3Aus-west-2%3A123456789012%3Akey%2F11111111-2222-3333-4444-555555555555" +
+                "&DestinationRegion=us-west-2" +
+                "&X-Amz-Algorithm=AWS4-HMAC-SHA256" +
+                "&X-Amz-Date=20161221T180735Z" +
+                "&X-Amz-SignedHeaders=host" +
+                "&X-Amz-Expires=604800" +
+                "&X-Amz-Credential=foo%2F20161221%2Fus-east-1%2Frds%2Faws4_request" +
+                "&X-Amz-Signature=00822ebbba95e2e6ac09112aa85621fbef060a596e3e1480f9f4ac61493e9821";
+        assertEquals(expectedPreSignedUrl, presignedRequest.rawQueryParameters().get("PreSignedUrl").get(0));
```
Comment (Quanzzzz) [inline services/neptune/src/test/java/software/amazon/awssdk/services/neptune/internal/PresignRequestHandlerTest.java:102]: Same as above, can we use the `equals` method here?

**Review by Quanzzzz — APPROVED** [review]

File: services/docdb/src/test/java/software/amazon/awssdk/services/docdb/internal/PresignRequestHandlerTest.java [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+                "&TargetDBClusterSnapshotIdentifier=test-instance-ss-copy-2" +
+                "&KmsKeyId=arn%3Aaws%3Akms%3Aus-west-2%3A123456789012%3Akey%2F11111111-2222-3333-4444-555555555555" +
+                "&DestinationRegion=us-west-2" +
+                "&X-Amz-Algorithm=AWS4-HMAC-SHA256" +
+                "&X-Amz-Date=20161221T180735Z" +
+                "&X-Amz-SignedHeaders=host" +
+                "&X-Amz-Expires=604800" +
+                "&X-Amz-Credential=foo%2F20161221%2Fus-east-1%2Frds%2Faws4_request" +
+                "&X-Amz-Signature=00822ebbba95e2e6ac09112aa85621fbef060a596e3e1480f9f4ac61493e9821";
+        assertEquals(expectedPreSignedUrl, presignedRequest.rawQueryParameters().get("PreSignedUrl").get(0));
```
[the remark this answers is not in the document]
Comment (joviegas) [reply services/docdb/src/test/java/software/amazon/awssdk/services/docdb/internal/PresignRequestHandlerTest.java:106]: We will need to create expected presignedRequest in that case.
The other attributes are not important w.r.t test case, so we are just asserting on rawQueryParameters which is updated by interceptors.

File: services/neptune/src/test/java/software/amazon/awssdk/services/neptune/internal/PresignRequestHandlerTest.java (hunk 5 above)
Comment (joviegas) [reply services/neptune/src/test/java/software/amazon/awssdk/services/neptune/internal/PresignRequestHandlerTest.java:102]: We will need to create expected presignedRequest in that case.
The other attributes are not important w.r.t test case, so we are just asserting on rawQueryParameters which is updated by interceptors

**Pull request merged**